frontend/packages/console-plugin-sdk/src/useExtensions.ts

```diff
@@ -0,0 +1,75 @@
+import { useMemo } from 'react';
+// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+// @ts-ignore: FIXME missing exports due to out-of-sync @types/react-redux version
+import { useSelector, shallowEqual } from 'react-redux';
+import { RootState } from '@console/internal/redux';
+import { FlagsObject } from '@console/internal/reducers/features';
+import { pluginStore } from '@console/internal/plugins';
+import { getFlagsFromState, isExtensionInUse } from './store';
+import { Extension, ExtensionTypeGuard } from './typings';
+
+/**
+ * React hook for consuming Console extensions.
+ *
+ * This hook takes an extension type guard as an argument and returns a list of
+ * extension instances (narrowed according to the type guard) which are currently
+ * in use.
+ *
+ * An extension is in use when
+ *
+ * - it is an always-on extension, i.e. not gated by feature flags
+ * - all feature flags referenced by its `flags` object are resolved to the right
+ *   values
+ *
+ * For example:
+ *
+ * ```ts
+ * import { isFooNavItem, isNavItem } from '@console/plugin-sdk';
+ *
+ * const Example = () => {
+ *   // filter by single extension type
+ *   // return type is FooNavItem[]
+ *   const fooNavItems = useExtensions(isFooNavItem);
+ *
+ *   // filter by union of extension types
+ *   // return type is NavItem[] = (FooNavItem | BarNavItem)[]
+ *   const allNavItems = useExtensions(isNavItem);
+ *
+ *   // process extensions and render component
+ * };
+ * ```
+ *
+ * You may call `useExtensions` multiple times in a single functional component
+ * to retrieve Console extensions of different types. Internally, this hook calls
+ * `useSelector` from `react-redux` library, i.e. each call creates an individual
+ * subscription to the Redux store.
+ *
+ * _For now, the runtime list of extensions is assumed to be immutable._
+ *
+ * @param typeGuard Type guard used to narrow the extension type.
+ */
+export const useExtensions: UseExtensions = (typeGuard) => {
+  const allExtensions = pluginStore.getAllExtensions();
+
+  // Select extensions according to type guard
+  const desiredExtensions = useMemo(() => allExtensions.filter(typeGuard), [
+    allExtensions,
+    typeGuard,
+  ]);
+
+  // Compute flags relevant for gating desired extensions
+  const desiredFlags: FlagsObject = useSelector(
```

Comment:
you can specify state and return value by
```suggestion
  const desiredFlags = useSelector<RootState, FlagsObject>(
```

Reply:
@rawagner Please see https://github.com/openshift/console/pull/3383#discussion_r346540121 - we're currently using an outdated version of `@types/react-redux` so we're missing types for new `react-redux` APIs like `useSelector` and `shallowEqual`.

I'll add type params to `useSelector` call as suggested.

However, we still need the explicit `desiredFlags: FlagsObject` type declaration due to missing types - otherwise, this variable would be inferred as `any`.